src/commands/classified/external.rs

```diff
@@ -50,164 +48,251 @@ impl Decoder for LinesCodec {
 
-    if let Some(stdin) = stdin {
-        process = process.stdin(stdin);
+    if !is_last {
+        process = process.stdout(subprocess::Redirection::Pipe);
+        trace!(target: "nu::run::external", "set up stdout pipe");
     }
 
+    process = process.stdin(subprocess::Redirection::Pipe);
     trace!(target: "nu::run::external", "set up stdin pipe");
+
     trace!(target: "nu::run::external", "built process {:?}", process);
 
-    let popen = process.popen();
+    let popen = process.detached().popen();
+    if let Ok(mut popen) = popen {
+        let mut stdin_write = popen.stdin.take().unwrap();
+
+        let stream = async_stream! {
+            while let Some(item) = input.next().await {
+                match item.value {
+                    UntaggedValue::Primitive(Primitive::Nothing) => {
+                        // If first in a pipeline, will receive Nothing. This is not an error.
+                    },
 
-    trace!(target: "nu::run::external", "next = {:?}", stream_next);
+                    UntaggedValue::Primitive(Primitive::String(s)) |
+                        UntaggedValue::Primitive(Primitive::Line(s)) =>
+                    {
+                        if let Err(e) = stdin_write.write(s.as_bytes()) {
+                            yield Ok(Value {
+                                value: UntaggedValue::Error(ShellError::labeled_error(
+                                    format!("Unable to write to stdin (error = {})", e),
+                                    "unable to write to stdin",
+                                    name_tag.clone(),
+                                )),
+                                tag: name_tag,
+                            });
+                            return;
+                        }
+                    },
 
-    let name_tag = command.name_tag.clone();
-    if let Ok(mut popen) = popen {
-        popen.detach();
-        match stream_next {
-            StreamNext::Last => {
-                let _ = popen.wait();
-                Ok(ClassifiedInputStream::new())
-            }
-            StreamNext::External => {
-                let stdout = popen.stdout.take().ok_or_else(|| {
-                    ShellError::untagged_runtime_error(
-                        "Can't redirect the stdout for external command",
-                    )
-                })?;
-                Ok(ClassifiedInputStream::from_stdout(stdout))
+                    // TODO perhaps serialize some other primitives?
+
+                    v => {
+                        yield Ok(Value {
+                            value: UntaggedValue::Error(ShellError::labeled_error(
+                                format!("Unexpected stdin type found in input stream ({})", v.type_name()),
+                                "unexpected type",
+                                name_tag.clone(),
+                            )),
+                            tag: name_tag,
+                        });
```

Comment:
There's a lot of these floating around. Is this the best way to do this?

Reply:
instead of cloning?  In some places you can borrow tags

Reply:
Just the whole pattern in general, but definitely reducing the clones would be 👌

Reply:
I think a follow-up PR to make this return an `OutputStream` will actually simplify things. I have some ideas to simplify this 🙂

Reply:
Looks like this: ![screenshot](https://user-images.githubusercontent.com/833576/71906403-00d35180-3138-11ea-9491-437b02d8dc4a.png)

Reply:
I might say something like "expected a string" for the label, but yeah that looks just about right

Reply:
Ah yes, that's probably more appropriate for now. If we ever did allow serializing other types in the future, we can adjust the message. Will make that change 👍